runtimes/pythonrt/worker_grpc_handler.go

```diff
@@ -257,140 +260,110 @@ func (s *workerGRPCHandler) StartSession(ctx context.Context, req *userCode.Star
+	if err != nil {
+		return &userCode.SubscribeResponse{Error: err.Error()}, nil
 	}
-	msg := makeCallbackMessage(args, nil)
-	runner.channels.callback <- msg
 
-	select {
-	case err := <-msg.errorChannel:
+	if resp.err != nil {
 		err = status.Errorf(codes.Internal, "subscribe(%s, %s) -> %s", req.Connection, req.Filter, err)
 		return &userCode.SubscribeResponse{Error: err.Error()}, nil
-	case val := <-msg.successChannel:
-		signalID := val.GetString().Value()
-		return &userCode.SubscribeResponse{SignalId: signalID}, nil
 	}
+
+	return &userCode.SubscribeResponse{SignalId: resp.value.(string)}, nil
 }
 
 func (s *workerGRPCHandler) NextEvent(ctx context.Context, req *userCode.NextEventRequest) (*userCode.NextEventResponse, error) {
 	if len(req.SignalIds) == 0 {
 		w.log.Error("empty signal ID")
 		return nil, status.Error(codes.InvalidArgument, "at least one signal ID required")
 	}
+
 	if req.TimeoutMs < 0 {
 		w.log.Error("bad timeout", zap.Int64("timeout", req.TimeoutMs))
 		return nil, status.Error(codes.InvalidArgument, "timeout < 0")
 	}
 
-	w.mu.Lock()
-	runner, ok := w.runnerIDsToRuntime[req.RunnerId]
-	w.mu.Unlock()
-	if !ok {
-		w.log.Error("unknown runner ID", zap.String("id", req.RunnerId))
-		return &userCode.NextEventResponse{Error: "unknown runner ID"}, nil
+	fn := func(ctx context.Context, cbs *sdkservices.RunCallbacks, rid sdktypes.RunID) (any, error) {
+		return cbs.SafeNextEvent(ctx, rid, req.SignalIds, time.Duration(req.TimeoutMs)*time.Millisecond)
 	}
 
-	args := make([]sdktypes.Value, len(req.SignalIds)+1)
-	args[0] = sdktypes.NewStringValue("next_event")
-	for i, id := range req.SignalIds {
-		args[i+1] = sdktypes.NewStringValue(id)
-	}
-	// timeout is kw only
-	kw := make(map[string]sdktypes.Value)
-	if req.TimeoutMs > 0 {
-		kw["timeout"] = sdktypes.NewFloatValue(float64(req.TimeoutMs) / 1000.0)
+	resp, err := s.callback(ctx, req.RunnerId, "next_event", fn)
+	if err != nil {
+		return &userCode.NextEventResponse{Error: err.Error()}, nil
 	}
 
-	msg := makeCallbackMessage(args, kw)
-	runner.channels.callback <- msg
-
-	select {
-	case err := <-msg.errorChannel:
+	if resp.err != nil {
 		err = status.Errorf(codes.Internal, "next_event(%s, %d) -> %s", req.SignalIds, req.TimeoutMs, err)
 		return &userCode.NextEventResponse{Error: err.Error()}, nil
-	case val := <-msg.successChannel:
-		out, err := sdktypes.ValueWrapper{SafeForJSON: true}.Unwrap(val)
-		if err != nil {
-			err = status.Errorf(codes.Internal, "can't unwrap %v - %s", val, err)
-			return &userCode.NextEventResponse{Error: err.Error()}, err
-		}
-
-		data, err := json.Marshal(out)
-		if err != nil {
-			err = status.Errorf(codes.Internal, "can't json.Marshal %v - %s", out, err)
-			return &userCode.NextEventResponse{Error: err.Error()}, err
-		}
-
-		resp := userCode.NextEventResponse{
-			Event: &userCode.Event{
-				Data: data,
-			},
-		}
-		return &resp, nil
 	}
+
+	out, err := sdktypes.ValueWrapper{SafeForJSON: true}.Unwrap(resp.value.(sdktypes.Value))
```

Comment:
The type assertion might panic, use ", ok" first.

Reply:
it can't - it's always an `sdktypes.Value` here.